// remark-mdx-frontmatter/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.remarkMdxFrontmatter = void 0;
const estree_util_is_identifier_name_1 = require("estree-util-is-identifier-name");
const estree_util_value_to_estree_1 = require("estree-util-value-to-estree");
const js_yaml_1 = require("js-yaml");
const toml_1 = require("toml");
const visit = require("unist-util-visit");
/**
 * A remark plugin to expose frontmatter data as named exports.
 *
 * @param options - Optional options to configure the output.
 * @returns A unified transformer.
 */
const remarkMdxFrontmatter = ({ name } = {}) => (ast) => {
    const imports = [];
    if (name && !estree_util_is_identifier_name_1.name(name)) {
        throw new Error(`If name is specified, this should be a valid identifier name, got: ${JSON.stringify(name)}`);
    }
    visit(ast, (node) => {
        let data;
        if (node.type === 'yaml') {
            data = js_yaml_1.load(node.value);
        }
        else if (node.type === 'toml') {
            data = toml_1.parse(node.value);
        }
        if (data == null) {
            return;
        }
        if (!name && typeof data !== 'object') {
            throw new Error(`Expected frontmatter data to be an object, got:\n${node.value}`);
        }
        imports.push({
            type: 'mdxjsEsm',
            data: {
                estree: {
                    type: 'Program',
                    sourceType: 'module',
                    body: [
                        {
                            type: 'ExportNamedDeclaration',
                            source: null,
                            specifiers: [],
                            declaration: {
                                type: 'VariableDeclaration',
                                kind: 'const',
                                declarations: Object.entries(name ? { [name]: data } : data).map(([identifier, value]) => {
                                    if (!estree_util_is_identifier_name_1.name(identifier)) {
                                        throw new Error(`Frontmatter keys should be valid identifiers, got: ${JSON.stringify(identifier)}`);
                                    }
                                    return {
                                        type: 'VariableDeclarator',
                                        id: { type: 'Identifier', name: identifier },
                                        init: estree_util_value_to_estree_1.valueToEstree(value),
                                    };
                                }),
                            },
                        },
                    ],
                },
            },
        });
    });
    ast.children.unshift(...imports);
};
exports.remarkMdxFrontmatter = remarkMdxFrontmatter;
